### ISSUE TITLE
seccomp does not support multiple arguments' rules working together

### ISSUE THREAD

**Issue by ihac** [moby/moby#32714]
@mheon @jessfraz @justincormack
Hi, all. I'm working on generating suitable seccomp profiles for any Docker container. However, I'm not quite content with the functionalities of Docker seccomp, which only support limiting syscall by its name or some argument.
I read the [default seccomp profile](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json) before, and I found that you put two argument's rules into one syscall object(e.g. [Line 525](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json#L525)). Seems that you were trying to let them work together(i.e. only ALLOW when those two rules are both matched). But actually it does not work that the syscall will be allowed when any one of the two rules is matched. I verified this by testing with some customized profiles.
I'm pretty sure that `libseccomp` supports applying multiple argument's rules together(refer to [my topic](https://groups.google.com/forum/#!topic/libseccomp/juKfwujESiQ)). So I checked the source code of `libseccomp-golang`, only found that it set one condition(argument's rule) per time([Line 310](https://github.com/moby/moby/blob/master/vendor/github.com/seccomp/libseccomp-golang/seccomp_internal.go#L310)). 

```
for _, cond := range conds {
	cmpStruct := C.make_struct_arg_cmp(C.uint(cond.Argument), cond.Op.toNative(), C.uint64_t(cond.Operand1), C.uint64_t(cond.Operand2))
	defer C.free(cmpStruct)

	if err := f.addRuleWrapper(call, action, exact, C.scmp_cast_t(cmpStruct)); err != nil {
		return err
	}
}
```
I was very confused with the code here. Why use `seccomp_rule_add_[exact_]array()` while only setting one rule(instead of multiple rules) a time, `seccomp_rule_add[_exact]()` could do the same. 

So, here are my questions:

1. Is it in your milestone that supporting multiple arguments(of one syscall)' rules AND'd together. If not, is this feature acceptable enough for creating a PR?

2. Is there any possible way for developers to set seccomp bpf filters for containers without using libseccomp interfaces? I'd like to limit app with more complex rules, such as limiting based on syscall sequences?

3. One more puzzle. I have difficulty in understanding the default seccomp profile. what's the meaning of the magic number `2080505856`(or `0x7C020000`) in `clone`'s rule([Line 765](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json#L765)). I've searched internet for a while but got thing.

Thanks for your attention. Looking forward to your reply.

**Output of `docker version`:**

```
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 3
Server Version: 17.04.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 24
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary:
containerd version: 422e31ce907fd9c3833a38d7b8fdd023e5a76e73
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-72-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.853GiB
Name: ubuntu-one
ID: KTSQ:V7MO:PC3E:QMPV:A3F7:4M7S:UAIQ:3JGK:GFM6:O4CD:LHDT:6OCU
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

KVM guest with Ubuntu 16.04.2 LTS installed.

**Comment by mheon** [comment]
I believe what you're describing sounds like a bug in `libseccomp` versions lower than v2.2.1. On higher library versions, adding per-argument rules one at a time will correctly generate the correct result (the match will be the AND of all conditions). Can you verify what version of `libseccomp` you have on your host?

**Comment by ihac** [comment]
@mheon 
Well, it's v2.2.3 if my program is correct:

```
root@xxx:~/libseccomp-test# cat version1.c
#include "stdio.h"
#include "seccomp.h"
int main() {
  printf("%d %d %d\n", SCMP_VER_MAJOR, SCMP_VER_MINOR, SCMP_VER_MICRO);
  return 0;
}
root@xxx:~/libseccomp-test# gcc version1.c -o version1 && ./version1
2 2 3
root@xxx:~/libseccomp-test# 
``` 

> On higher library versions, adding per-argument rules one at a time will correctly generate the correct result (the match will be the AND of all conditions).

I doubt about it since I've tried both `libseccomp` API and Docker seccomp profiles for applying multiple argument's rules but failed. 
Meanwhile, multiple `seccomp_rule_add[_xxx]()` rules are OR'd together according to [Paul](https://groups.google.com/d/msg/libseccomp/juKfwujESiQ/aG3AJ_sjEwAJ), 

> At present all of the argument rules in a single seccomp_rule_add() 
call are AND'd together, and multiple seccomp_rule_add() rules are 
OR'd together.

**Comment by ihac** [comment]
FYI, I used some never-matched rules in the test profile:

```
...
		{
			"names": [
				"fcntl"
			],
			"action": "SCMP_ACT_ALLOW",
			"args": [
				{
					"index": 0,
					"value": 10,
					"valueTwo": 0,
					"op": "SCMP_CMP_LE"
				},
				{
					"index": 1,
					"value": 0,
					"valueTwo": 0,
					"op": "SCMP_CMP_LT"
				}
			],
			"comment": "",
			"includes": {},
			"excludes": {}
		},
...
```
Obviously, the second rule could never be matched. If the two rules are AND'd together, `fcntl()` should not be allowed. But in fact, it's allowed.

**Comment by mheon** [comment]
Hm. Looks like you're correct. In this case, `libseccomp` does not appear to be able to express what you're trying to do. You'll need arbitrary loading of BPF filters to do this. Unfortunately, this isn't available today.

There has been discussion about allowing arbitrary BPF to be loaded, but that work hasn't made it into the OCI runtime spec yet, and as such is not supported. If it does make it into the spec, `libseccomp` itself does support loading arbitrary BPF filters, so it should be easy enough to add support to `runc` once it's in the spec.

**Comment by ihac** [comment]
So I could never load arbitrary BPF filters into kernel for now if i understand correctly.
However, `libseccomp` itself does support `AND` for multiple per-argument rules.
E.g.

```
 seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(some_syscall), 2, 
        SCMP_A0(SCMP_CMP_LE, 2), 
        SCMP_A1(SCMP_CMP_GE, 1) 
  ); 
```
The code above works well, two rules(for **different arguments** of **one syscall**) are AND'd together.

I am pretty sure it's `libseccomp-golang` that omits the functionality, since it only loads one rule per `seccomp_rule_add_[exact_]array()`.

**Comment by mheon** [comment]
After playing around with this a bit, it seems like you're right. Pretty glaring error on my part (oops!). I'm working on some patches for `libseccomp-golang` already; I'll see about getting a fix for this added to them. Problem is that our current default profile assumes that conditional matches are OR based, and changing matching to AND would break the default filter (and probably user-generated filters as well). I'm still going to change the bindings (the current behavior does not match the library and is definitely wrong because of this), but I'll need to figure out a way to preserve existing filters while doing so.

**Comment by ihac** [comment]
That's great for you to work on it!
Preserving existing filters is vital, but not when they did not work as expected. 
E.g. [Line 501](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json#L503) of the default profile.
```
{
  "names": [
    "socketcall"
  ],
  "action": "SCMP_ACT_ALLOW",
  "args": [
    {
      "index": 0,
      "value": 1,
      "valueTwo": 0,
      "op": "SCMP_CMP_GT"
    }
  ],
  "comment": "",
  "includes": {},
  "excludes": {}
},
{
  "names": [
    "socketcall"
  ],
  "action": "SCMP_ACT_ALLOW",
  "args": [
    {
      "index": 0,
      "value": 1,
      "valueTwo": 0,
      "op": "SCMP_CMP_EQ"
    },
    {
      "index": 1,
      "value": 1,
      "valueTwo": 0,
      "op": "SCMP_CMP_EQ"
    }
  ],
  "comment": "",
  "includes": {},
  "excludes": {}
},
```
Seems that they were trying to limit `socketcall()` so that it would be allowed only when, the `0th` argument is greater than `1`(i.e. not `socket()`), **OR** the `0th` argument is equal to `1`(i.e. `socket`) **and** `1th` argument is equal to `1`(maybe `1` here refers to `AF_LOCAL`?), if I am not mistaken. (btw, @justincormack , I guess it's wrong to compare the second argument of `socketcall()` with a number `1`, since it's a pointer)
In short, three conditions are not 'OR'd together, but `cond1 OR (cond2 AND cond3)`. If so, the current default profile does not work as expected for now. It won't break the expected behaviour of the default filter/user-generated filters, if you just load the multiple rules in the same `args` list with calling `seccomp_rule_add()` exactly once.

**Comment by justincormack** [comment]
@ihac the `socket` rules have been removed due to the pointer issue....

**Comment by justincormack** [comment]
I can't see any rules now that have multiple arguments, but will update the library.
